waiter/src/waiter/auth/composite.clj

```diff
@@ -0,0 +1,69 @@
+            [waiter.util.utils :as utils]))
+
+(defrecord CompositeAuthenticator [provider-name->authenticator default-authentication-provider]
+  auth/Authenticator
+
+  (get-authentication-providers [_]
+    (keys provider-name->authenticator))
+  (process-callback [_ {{:keys [authentication-provider]} :route-params :as request}]
+    (if-let [authenticator (get provider-name->authenticator authentication-provider)]
+      (auth/process-callback authenticator request)
+      (throw (ex-info (str "Unknown authentication provider " authentication-provider)
+                      {:available-authenticators (keys provider-name->authenticator)
+                       :authentication-provider authentication-provider
+                       :status 400}))))
+  (wrap-auth-handler [_ request-handler]
+    (let [provider-name->handler (pc/map-vals #(auth/wrap-auth-handler % request-handler) provider-name->authenticator)]
+      (fn composite-authenticator-handler [request]
+        (let [authentication (or (get-in request [:waiter-discovery :service-parameter-template "authentication"])
+                                 default-authentication-provider)]
+          (if-let [handler (get provider-name->handler authentication)]
+            (handler request)
+            (throw (ex-info (str "No authenticator found for " authentication " authentication.")
+                            {:available-authenticators (keys provider-name->handler)
+                             :request-authentication authentication
+                             :status 400}))))))))
+
+(defn- make-authenticator
+  "Create an authenticator from an authentication-provider"
+  [{:keys [factory-fn] :as authentication-provider} context]
+  {:pre [(symbol? factory-fn)]}
+  (let [resolved-factory-fn (utils/resolve-symbol! factory-fn)
+        authenticator (resolved-factory-fn (merge context authentication-provider))]
+    (when-not (satisfies? auth/Authenticator authenticator)
+      (throw (ex-info "Authenticator factory did not create an instance of Authenticator"
+                      {:authentication-provider authentication-provider
+                       :authenticator authenticator
+                       :resolved-factory-fn resolved-factory-fn})))
+    authenticator))
+
+(defn composite-authenticator
+  "Factory function for creating composite authenticator middleware"
+  [{:keys [authentication-providers default-authentication-provider] :as context}]
+  {:pre [(not-empty authentication-providers)
+         (not (string/blank? default-authentication-provider))
+         (contains? authentication-providers default-authentication-provider)]}
+  (let [provider-name->authenticator (as-> (pc/map-vals
+                                             #(make-authenticator % context)
+                                             authentication-providers) map
+                                       (merge {"standard" (get map default-authentication-provider)} map))]
```

Comment:
I feel `waiter.util.utils/assoc-if-absent` would be easier to read than the `as->` and then `merge`.

Reply:
i'm doing a get from "map"

what's the code you're proposing?

Reply:
```
(assoc-if-absent (pc/map-vals #(make-authenticator % context) authentication-providers)
  "standard" (get map default-authentication-provider))
```

Reply:
I'm thinking about whether we need to add `standard` separately. If all the other authenticators provide `standard` in their default implementation (from the `extend-protocol`), then composite can just publish its list from the keys. If standard is not one of the keys, then it is not supported for the composite scheduler. That also simplifies our documentation to `disabled` or any supported authentication provider.

Reply:
> ```
> (assoc-if-absent (pc/map-vals #(make-authenticator % context) authentication-providers)
>   "standard" (get map default-authentication-provider))
> ```
"map" is not defined in this snippet

Reply:
I don't see a problem with the composite authenticator saying that it supports "standard" and it having the "standard" key in its map, and it just returning the keys of its map

what would the simplified documentation say?

Reply:
> "map" is not defined in this snippet

That is the point, it does not need it.

Reply:
> I don't see a problem with the composite authenticator saying that it supports "standard" and it having the "standard" key in its map, and it just returning the keys of its map

Extra code we don't need if we configure it right with one of the components being `standard`.

Reply:
> > "map" is not defined in this snippet
> 
> That is the point, it does not need it.

I don't understand. This code doesn't compile

```
(assoc-if-absent (pc/map-vals #(make-authenticator % context) authentication-providers)
  "standard" (get map default-authentication-provider))
```

Reply:
> Extra code we don't need if we configure it right with one of the components being `standard`.

You want the configuration of the composite authenticator to have an extra `standard` component in the config? If so, I don't like that

Reply:
> This code doesn't compile

oops scratch that, didn't notice the `map` in the `default-authentication-provider`.